runtime/gc_glue_java/MetronomeDelegate.cpp

```diff
@@ -1115,5 +1131,430 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+		J9Object *referent = referentSlotObject.readReferenceFromSlot();
+		if (NULL != referent) {
+			UDATA referenceObjectType = J9CLASS_FLAGS(J9GC_J9OBJECT_CLAZZ(referenceObj)) & J9AccClassReferenceMask;
+			if (_realtimeGC->_markingScheme->isMarked(referent)) {
+				if (J9AccClassReferenceSoft == referenceObjectType) {
+					U_32 age = J9GC_J9VMJAVALANGSOFTREFERENCE_AGE(env, referenceObj);
+					if (age < _extensions->getMaxSoftReferenceAge()) {
+						/* Soft reference hasn't aged sufficiently yet - increment the age */
+						J9GC_J9VMJAVALANGSOFTREFERENCE_AGE(env, referenceObj) = age + 1;
+					}
+				}
+			} else {
+				/* transition the state to cleared */
+				Assert_MM_true(GC_ObjectModel::REF_STATE_INITIAL == J9GC_J9VMJAVALANGREFERENCE_STATE(env, referenceObj));
+				J9GC_J9VMJAVALANGREFERENCE_STATE(env, referenceObj) = GC_ObjectModel::REF_STATE_CLEARED;
+
+				referenceStats->_cleared += 1;
+
+				/* Phantom references keep it's referent alive in Java 8 and doesn't in Java 9 and later */
+				if ((J9AccClassReferencePhantom == referenceObjectType) && ((J2SE_VERSION(_javaVM) & J2SE_VERSION_MASK) <= J2SE_18)) {
+					/* Scanning will be done after the enqueuing */
+					_realtimeGC->getMarkingScheme()->markObject(env, referent);
+				} else {
+					referentSlotObject.writeReferenceToSlot(NULL);
+				}
+#if defined(J9VM_GC_FINALIZATION)
+				/* Check if the reference has a queue */
+				if (0 != J9GC_J9VMJAVALANGREFERENCE_QUEUE(env, referenceObj)) {
+					/* Reference object can be enqueued onto the finalizable list */
+					buffer.add(env, referenceObj);
+					referenceStats->_enqueued += 1;
+					/* Flag for the finalizer */
+					_finalizationRequired = true;
+				}
+#endif /* J9VM_GC_FINALIZATION */
+			}
+		}
+		referenceObj = nextReferenceObj;
+		if (REFERENCE_OBJECT_YIELD_CHECK_INTERVAL == objectsVisited) {
+			_scheduler->condYieldFromGC(env);
+			objectsVisited = 0;
+		}
+	}
+#if defined(J9VM_GC_FINALIZATION)
+	buffer.flush(env);
+#endif /* J9VM_GC_FINALIZATION */
+}
+
+/**
+ * Mark all of the roots. The system and application classloaders need to be set
+ * to marked/scanned before root marking begins.
+ *
+ * @note Once the root lists all have barriers this code may change to call rootScanner.scanRoots();
+ * 
+ */
+void 
+MM_MetronomeDelegate::markRoots(MM_EnvironmentRealtime *env, MM_RealtimeMarkingSchemeRootMarker *rootScanner)
+{
+	/* Mark root set classes */
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+	if(env->isMasterThread()) {
+		/* TODO: This code belongs somewhere else? */
+		/* Setting the permanent class loaders to scanned without a locked operation is safe
+		 * Class loaders will not be rescanned until a thread synchronize is executed
+		 */
+		if(isDynamicClassUnloadingEnabled()) {
+			((J9ClassLoader *)_javaVM->systemClassLoader)->gcFlags |= J9_GC_CLASS_LOADER_SCANNED;
+			_realtimeGC->getMarkingScheme()->markObject(env, (J9Object *)((J9ClassLoader *)_javaVM->systemClassLoader)->classLoaderObject);
```

Comment:
`_realtimeGC->getMarkingScheme()` can be replaced to `_markingScheme` everywhere in `MM_MetronomeDelegate`